### PR TITLE
Support Rails 6.1 ActiveModel::Errors

### PR DESCRIPTION
**Pull request by rewritten**
ActiveModel::Errors class has been revamped and it is now not compatible with most of the existing (and already deprecated) idioms (adding errors with `record.errors[:attr] << 'My error'`, mutating the hash returned by `#messages`, and a big chunk of the remaining API is deprecated.

This change aims to support both pre-6.1 and post 6.1 versions with as little conditionals as possible.

It also includes a patch to ActiveData's nested validator, which attempts to mutate directly the error hash (which is not a hash anymore). A similar patch is available at pyromaniac/active_data#76.

### Review

- [ ] Document code according to [Getting Started with Yard](http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md).
- [ ] All tests are passing.
- [ ] Test manually.
- [ ] Get approval.

### Pre-merge checklist

- [ ] The PR relates to a single subject with a clear title and description in grammatically correct, complete sentences.
- [ ] Verify that feature branch is up-to-date with `master` (if not - rebase it).
- [ ] Double check the quality of [commit messages](http://chris.beams.io/posts/git-commit/).
- [ ] Squash related commits together.

### CODE REVIEW

File: spec/lib/granite/rspec/raise_validation_error_spec.rb [hunk 1]
```diff
@@ -29,27 +29,7 @@ def execute_perform!(*)
...
-      expect do
-        expect { action.perform! }.to raise_validation_error.on_attribute(:raise_error)
-      end.to fail_with('expected to raise validation error on attribute :raise_error, but raised {:base=>[{:error=>:some_error}], :raise_error=>[]}')
-    end
-
-    specify do
-      expect do
-        expect { action.perform! }.to raise_validation_error.on_attribute(:raise_error).of_type(:some_error)
-      end.to fail_with('expected to raise validation error on attribute :raise_error of type :some_error, but raised {:base=>[{:error=>:some_error}], :raise_error=>[]}')
+      expect { action.perform! }.to raise_validation_error.on_attribute(:base).of_type(:some_error)
```
Comment (rewritten) [inline spec/lib/granite/rspec/raise_validation_error_spec.rb:32]: I am not sure about the rationale behind those tests, it seems to me that a single expectation is enough to cover it.

**Comment by rewritten** [comment]
Coverage fails expectedly, because one of the branches in `lib/granite/action/preconditions/embedded_precondition.rb` is never used…

File: lib/granite/action/preconditions/embedded_precondition.rb [hunk 2]
```diff
@@ -34,9 +34,20 @@ def _execute(context)
...
         end
+
+        def merge_errors(target, source)
+          if source.respond_to?(:where)
+            # ActiveModel::Errors from version 6.1
+            source.where(:base).each { |error| target.import(error) }
+          else
+            # ActiveModel::Errors until version 6.0
+            source[:base].each { |error| target.add(:base, error) }
+          end
```
Comment (rewritten) [inline lib/granite/action/preconditions/embedded_precondition.rb:49]:
```suggestion
          source.respond_to?(:where) ? source.where(:base).each { |error| target.import(error) } : source[:base].each { |error| target.add(:base, error) }
```

File: spec/lib/granite/rspec/raise_validation_error_spec.rb (hunk 1 above)
Comment (rewritten) [reply spec/lib/granite/rspec/raise_validation_error_spec.rb]: Nevermind, I restored them

**Comment by rewritten** [comment]
Do you need to keep compatibility with ruby 2.4?

File: lib/granite/action.rb [hunk 3]
```diff
@@ -50,20 +50,28 @@ def assign_attributes(attributes)
...
+          (this + other).uniq
+        end
+      end
+    else
+      def merge_errors(other_errors)
+        other_errors.each do |error|
+          errors.import(error) unless errors.added?(error.attribute, error)
+        end
       end
     end
```
Comment (rewritten) [inline lib/granite/action.rb:76]: This branching is the reason coverage has decreased. Only one of the two branches will ever be used in a single test run.

Anyway, the tests run in ruby 2.4 use activemodel 5, so the legacy execution branch is covered.

File: spec/lib/granite/rspec/raise_validation_error_spec.rb [hunk 4]
```diff
@@ -43,13 +43,13 @@ def execute_perform!(*)
     specify do
       expect do
         expect { action.perform! }.to raise_validation_error.on_attribute(:raise_error)
-      end.to fail_with('expected to raise validation error on attribute :raise_error, but raised {:base=>[{:error=>:some_error}], :raise_error=>[]}')
+      end.to fail_with(include 'expected to raise validation error on attribute :raise_error, but raised {:base=>[{:error=>:some_error}]')
```
Comment (rewritten) [inline spec/lib/granite/rspec/raise_validation_error_spec.rb:46]: The extra reported error was an artifact due to older activemodel initializing a message array on access (and our usage of `deep_merge!`). Test is changed to only check the relevant part of the message.

File: Gemfile [hunk 5]
```diff
@@ -1,3 +1,22 @@
 source 'https://rubygems.org'
+git_source(:toptal) { |repo| "https://github.com/#{repo}.git" }
 
 gemspec
+
+gem 'active_data', github: 'toptal/active_data', branch: 'master'
+
+gem 'activerecord', '>= 5.0', '< 7'
```
Comment (jodosha) [inline Gemfile:8]: @rewritten Why this change?

Having dev dependencies in gemspec has the benefit of being compatible with the rest of the ecosystem. E.g. Rubygems maps reverse deps for a gem, by analysing gemspec files.

Unless there is a valid reason for this, I would ask you to revert this change. Thanks. 😄

**Review by jodosha — COMMENTED** [review]
@rewritten Code LGTM. Could you please clarify the gemspec / Gemfile question? Thanks 😄

File: Gemfile (hunk 5 above)
Comment (rewritten) [reply Gemfile:8]: Ok. But rubygems never uses development dependencies when resolving transitive deps. I don't know where else the development dependencies are used (except for rubygems website).

Comment (rewritten) [reply Gemfile]: For reference, Marc-André answer to the same topic in [What are the practical advantages of using `add_development_dependency` in .gemspec?](https://stackoverflow.com/a/62737144/384417)


> It's not clear there is any.
>
> In theory, RubyGems has the information necessary to run tests with add_development_dependency and test_files (see this question).
>
> Some believe that this should be outside the scope of RubyGems (see this huge thread).
>
> In practice, there is currently no real advantage, and the fact that RubyGems still installs test files by default is a drawback, as might be the lack of flexibility that Gemfile offers.

**Review by jodosha — APPROVED** [review]
@rewritten Thanks for all the patience. You did an excellent job. 👍

File: granite.gemspec [hunk 6]
```diff
@@ -13,12 +13,12 @@ Gem::Specification.new do |s|
   s.files       = `git ls-files`.split("\n").grep(/\A(app|lib|config|LICENSE)/)
   s.license     = 'MIT'
 
-  s.add_runtime_dependency 'actionpack', '>= 5.1', '< 6.1'
+  s.add_runtime_dependency 'actionpack', '>= 5.1', '< 7'
```
Comment (ojab) [inline granite.gemspec:16]: Looks bad without check for rails+active_data versions :/
Right now granite+rails-6.1 would just fail in unexpected places without knowledge that active_data from some-non-upstream git repo should be used.
